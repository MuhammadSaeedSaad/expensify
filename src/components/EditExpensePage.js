import React from "react";


const EditExpensePage = (props) => (
    <div>
        Editing the expense of id: {props.match.params.id}
    </div>
);
3
export default EditExpensePage;