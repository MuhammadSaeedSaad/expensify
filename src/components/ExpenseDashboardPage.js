import React from "react";

// main page
const ExpenseDashboardPage = () => (
    <div>
        this is from my dashboard component
    </div>
);

export default ExpenseDashboardPage;