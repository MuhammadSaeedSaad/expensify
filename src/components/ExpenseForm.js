import React from "react";
import moment from "moment";
import 'react-dates/initialize';
import { SingleDatePicker } from "react-dates";


export default class ExpenseForm extends React.Component {
    constructor (props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : "",
            amount: props.expense ? (props.expense.amount / 100).toString() : "",
            note: props.expense ? props.expense.note : "",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ""
        };
    
    }
    // we defined the function this way to avoid losing the (this binding)
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {      // "!amount ||" to allow the user to delete the value
            this.setState(() => ({ amount }))
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        // this.setState(() => ({ note: e.target.value }));
        // this way is better than the commented one .. look vid (11.8)
        this.setState(() => ({ note: e.target.value }));
    }

    onDateChange = (createdAt) => {
        if (createdAt) {    // to prevent the user from deleting the value in the datebox
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "please provide description and amount" }));
        } else {
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render () {
        return (
            <div>
                {this.state.error && <h4>{this.state.error}</h4>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" placeholder="description" value={this.state.description} onChange={this.onDescriptionChange} />
                    <input type="text" placeholder="amount" value={this.state.amount} onChange={this.onAmountChange} />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea placeholder="add a note for your expense (optional)" value={this.state.note} onChange={this.onNoteChange} ></textarea>
                    <button type="submit">Add Expense</button>
                </form>
            </div>
        );
    }
}