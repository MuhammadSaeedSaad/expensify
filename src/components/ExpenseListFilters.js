import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }))
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    onSortChange = (e) => {
        console.log("e.target.value: " + e.target.value)
        if (e.target.value === "date") {
            console.log("inside if value: " + e.target.value)
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            console.log("inside if value: " + e.target.value)
            this.props.sortByAmount();
        }
    }

    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);