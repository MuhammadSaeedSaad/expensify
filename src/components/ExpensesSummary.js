import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import getVisibleExpenses from "../selectors/expenses";
import getExpensesTotal from "../selectors/expenses-total";


export const ExpensesSummary = (props) => (
    <div>
        { 
            props.expenses.length === 1 ?
            (<h1>Viewing 1 expense total {getExpensesTotal(props.expenses)}</h1>) : 
            (<h1>Viewing {props.expenses.length} expenses total is {numeral(getExpensesTotal(props.expenses) / 100).format("$0,0.00")}</h1>)
        }
    </div>
);

const mapStateToDispatch = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToDispatch)(ExpensesSummary);