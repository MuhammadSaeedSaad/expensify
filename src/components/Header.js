import React from "react";
import { NavLink } from "react-router-dom";

// <Link> and <NavLink> from vid 9.6
const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to="/" exact={true} activeClassName="is-active">Home</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create Expense</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
    </header>
);

export default Header;