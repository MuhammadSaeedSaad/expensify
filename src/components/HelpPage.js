import React from "react";

const HelpPage = () => (
    <div>
        this is from the help page
    </div>
);

export default HelpPage;