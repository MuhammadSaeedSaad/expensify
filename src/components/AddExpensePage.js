import React from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from "../actions/expenses";
import { connect } from "react-redux";

const AddExpensePage = (props) => (
    <div>
        this is from my add expense component
        <ExpenseForm onSubmit={(expense) => { 
            props.dispatch(addExpense(expense));
            props.history.push("/");
        }} />
    </div>
);

export default connect()(AddExpensePage);
