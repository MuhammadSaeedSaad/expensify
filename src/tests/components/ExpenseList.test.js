import React from "react";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";
import { shallow } from "enzyme";

test("should test ExpenseList with array of Expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test("should test ExpenseList with an empty array", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});