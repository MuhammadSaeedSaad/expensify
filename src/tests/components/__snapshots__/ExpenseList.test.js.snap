// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ExpenseList with an empty array 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should test ExpenseList with array of Expenses 1`] = `
<div>
  <ExpenseListItem
    amount={100}
    createdAt={259200000}
    description="Rent"
    id="1"
    key="1"
    note="some note 1"
  />
  <ExpenseListItem
    amount={200}
    createdAt={0}
    description="german books"
    id="2"
    key="2"
    note="some note 2"
  />
  <ExpenseListItem
    amount={300}
    createdAt={-259200000}
    description="T books"
    id="3"
    key="3"
    note="some note 3"
  />
</div>
`;
